### PR TITLE
ContributionFlow: fix previous button functionality in payment step

### PR DESCRIPTION
**Pull request by flickz**
This PR adds `stepsPayment` state to `Steps` component uses the information to skip validation when the user attempts to goBack or jump to previous step from payment step.

#### Screen record. 

Peek 2019-05-04 16-51

Ref([#1970](https://github.com/opencollective/opencollective/issues/1970))

### CODE REVIEW

File: src/components/Steps.js [hunk 1]
```diff
@@ -165,8 +167,13 @@ export default class Steps extends React.Component {
       return false;
     }
 
+    const goToStepOpt = {};
+    const stepPayment = this.props.stepPayment || {};
     const prevStep = this.props.steps[currentStep.index - 1];
-    this.goToStep(this.buildStep(prevStep, currentStep.index - 1));
+    if (currentStep.isLastStep && stepPayment.isNew && (stepPayment.error || !stepPayment.data)) {
```
Comment (znarf) [inline src/components/Steps.js]: Would be great to add code comments to explain in plain english what we're doing here.

File: src/components/Steps.js [hunk 2]
```diff
@@ -175,6 +182,12 @@ export default class Steps extends React.Component {
    * if `opts.ignoreValidation` is true.
    */
   goToStep = async (step, opts = {}) => {
+    const currentStep = this.getStepByName(this.props.currentStepName);
+    const stepPayment = this.props.stepPayment || {};
+    if (currentStep.isLastStep && stepPayment.isNew && (stepPayment.error || !stepPayment.data)) {
```
Comment (znarf) [inline src/components/Steps.js]: Same as before, would be great to add code comments to explain in plain english what we're doing here.

**Comment by znarf** [comment]
Looks good to me. Waiting on @Betree's review.

**Comment by Betree** [comment]
@flickz Something that may help:

In `Steps`, we can pass an `action` parameter to `currentStep.validate()` to indicate what triggered the validation. For example `goNext` can pass `next` and `goBack` `prev`.

Then in step payment's `validate`, we can do something like:

```es6
validateStepPayment = async (action) => {
  // No need to validate when going back
  if (action === 'prev') {
    return true
  }

  ...
}
```

**Comment by flickz** [comment]
@Betree Just a quick clarification, do you suggest we ignore validation completely when going back from payment step regardless if `CardInput` is empty or not?

**Comment by Betree** [comment]
@flickz I'm not, my example was a simplification just to demonstrate the change with steps but you can still use the condition that you previously added.

**Comment by flickz** [comment]
@Betree Can you check this? Have updated it since last Friday.

**Comment by znarf** [comment]
@flickz thank you! please don't abuse of force push. It's great to be able to look at follow up commits addressing comments from review.

File: src/components/Steps.js [hunk 3]
```diff
@@ -175,7 +175,12 @@ export default class Steps extends React.Component {
    * if `opts.ignoreValidation` is true.
    */
   goToStep = async (step, opts = {}) => {
-    if (!opts.ignoreValidation && !(await this.validateCurrentStep())) {
+    const currentStep = this.getStepByName(this.props.currentStepName);
+    if (currentStep.isLastStep && step.index < currentStep.index) {
+      opts.action = 'prev';
```
Comment (Betree) [inline src/components/Steps.js]: Why only for last step? Going back from 2nd step to first step is going `prev` too

Comment (flickz) [reply src/components/Steps.js]: I focused on the payment step which I assumed was the `lastStep` in every case. I didn't want other step validation behaviors to be altered.

Comment (Betree) [reply src/components/Steps.js]: It's a new parameter so others `validate` functions should be safe. From an API consumer point of view, one should expect that `action` is always `prev` when going back, regardless of whether this is the last step or not. Remember that `Steps` component is meant to be generic, so we should treat it as if it were an open source NPM package meant for public use :slightly_smiling_face:

**Comment by flickz** [comment]
I think this is good to go now.
